internal/inputs/inputs.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inputs
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/google/gke-policy-automation/internal/log"
+)
+
+const (
+	defaultMaxDataGetCoroutines = 20
+)
+
+type Input interface {
+	GetID() string
+	GetDescription() string
+	GetData(clusterID string) (interface{}, error)
+	Close() error
+}
+
+type Cluster struct {
+	Data map[string]interface{}
+}
+
+type getDataTask struct {
+	input     Input
+	clusterID string
+}
+
+type getDataTaskResult struct {
+	clusterID string
+	inputID   string
+	result    interface{}
+	err       error
+}
+
+//GetAllInputsData fetches data from given inputs for all given clusters in a concurrent manner
+func GetAllInputsData(inputs []Input, clusterIDs []string) (map[string]*Cluster, []error) {
+	tasksNo := len(inputs) * len(clusterIDs)
+	tasksChan := make(chan *getDataTask, tasksNo)
+	resultsChan := make(chan *getDataTaskResult, tasksNo)
+	errorsChan := make(chan *getDataTaskResult, tasksNo)
+	log.Infof("creating %d get data tasks", tasksNo)
+
+	for _, input := range inputs {
+		for _, clusterID := range clusterIDs {
+			tasksChan <- &getDataTask{input: input, clusterID: clusterID}
+		}
+	}
+	close(tasksChan)
+
+	log.Debugf("starting %d goroutines", defaultMaxDataGetCoroutines)
+	var wg sync.WaitGroup
+	for i := 0; i < defaultMaxDataGetCoroutines; i++ {
+		wg.Add(1)
+		go getInputData(i, &wg, tasksChan, resultsChan, errorsChan)
+	}
+	wg.Wait()
+	close(resultsChan)
+	close(errorsChan)
+	log.Debugf("processing results and errors")
+	results := processResults(resultsChan)
+	errors := processErrors(errorsChan)
+	return results, errors
+}
+
+func getInputData(i int, wg *sync.WaitGroup, tasks chan *getDataTask, results chan *getDataTaskResult, errors chan *getDataTaskResult) {
+	defer wg.Done()
```

Comment:
why defer here instead of calling Done at end?

Reply:
In this case the effect will be the same (function has no return statements).
However using defer in golang is a nice habit, as it always executes when the function returns. In this case it doesnt matter that much, but is useful i..e when the function has multiple return statements.